### ISSUE TITLE
docker has deprectate the device mapper storage driver, service fails to start now on zfs systems

### ISSUE THREAD

**Issue by con-f-use** [NixOS/nixpkgs#268552]
### Describe the bug
ZFS does not (yet) support overlayfs. Overlayfs is the default storage driver for the docker daemon. Docker has recently deprecated the old [devicemapper storage driver](https://github.com/moby/moby/pull/43378) and made it mandatory to specify that explicitly, your you'll get this message and the daemon fails to start:

```shell
$ journalctl --no-pager -eu docker.service
...
dockerd[335293]: failed to start daemon: error initializing graphdriver: prior storage driver devicemapper is deprecated and will be removed in a future release; update the the daemon configuration and explicitly choose this storage driver to continue using it; visit https://docs.docker.com/go/storage-driver/ for more information
```

### Steps To Reproduce
Happens with latest unstable (c757e9bd77b16ca2e03c89bf8bc9ecb28e0c06ad) if `/var/lib/docker` is located on a zfs mount and `virtualisation.docker.storageDriver` is the default `null`.

### Expected behavior

For ZFS, the default should probably be the `virtualisation.docker.storageDriver = "devicemapper"` until overlayfs-support will arrive in ZFS 2.2 (as announced).

**Comment by con-f-use** [comment]
@offlinehacker @tailhook @vdemeester @periklis @mikroskeem @maxbrunet

**Comment by vdemeester** [comment]
Feels like the deprecation of it dates of 2018. The “failure, or forcing to explicitly set it is more recent. 

How is you docker configuration in nixos ?
I think the only workaround here is to use the modules options to explicitly set the storage driver to “devicemapper” on your setup to make the docker daemon able to start or explicitly using the zfs device driver ? (https://docs.docker.com/storage/storagedriver/zfs-driver/#configure-docker-with-the-zfs-storage-driver)

**Comment by ncfavier** [comment]
Regarding the mass pings on your recent PRs, @con-f-use, it seems like you are using git incorrectly: checking out your PR https://github.com/NixOS/nixpkgs/pull/268681 and running `git log`, I see *exactly one* commit with an enormous diff, which seems to indicate that you've somehow squashed *the entire history* of nixpkgs into a single commit. You can see this with `git show` before pushing.

What `git` commands are you running?

**Comment by con-f-use** [comment]
Yeah, I'm terribly sorry, I did that exact thing numerous times before, without it ever exploding in my face like that.

I was editing the file directly on github, no git-cli involved. Just searched the file in `NixOs/nixpkgs`, clicked edit, made my changes, clicked commit, it would ask me if I'd wanted to fork, said yes, then I'd click "Make PR".

In the case of https://github.com/NixOS/nixpkgs/pull/268681, when I was asked to edit & reformat my commit message, I did a `git clone --branch con-f-use-patch-1 --depth 1 git@github.com:con-f-use/nixpkgs.git && cd nixpkgs && git commit --amend && git push origin con-f-use-patch-1 --force`

**Comment by ncfavier** [comment]
> `--depth 1`

That's your culprit. You should clone a full copy of nixpkgs, including the history, if you're going to work on it locally.

**Comment by con-f-use** [comment]
> Feels like the deprecation of it dates of 2018. The “failure, or forcing to explicitly set it is more recent.
> 
> How is you docker configuration in nixos ? I think the only workaround here is to use the modules options to explicitly set the storage driver to “devicemapper” on your setup to make the docker daemon able to start or explicitly using the zfs device driver ? (https://docs.docker.com/storage/storagedriver/zfs-driver/#configure-docker-with-the-zfs-storage-driver)

@vdemeester 

That's my configuration, nothing too special:

```nix
      virtualisation.docker = {
        enable = true;
        enableOnBoot = true;
        autoPrune.enable = true;
        daemon.settings = {
          insecure-registries = [
             # ...
          ];
          dns = ...;
        };
      };
```

I added `storageDriver = "zfs"` on that particular system, and made sure I had [backed-up images and containers](https://docs.docker.com/storage/storagedriver/select-storage-driver/#check-your-current-storage-driver) I needed.

**Comment by con-f-use** [comment]
Re my pull requests, seems to happen not only to me: https://github.com/NixOS/nixpkgs/pull/267272

**Comment by vdemeester** [comment]
> I wonder why docker used `devicemapper` in the first place, it should try a bunch of storage drivers and the first one working. Which should be "zfs" anyway, without the need for specifying it. Something is not having sensible defaults here, no idea if it's the nixos-module or docker itself.

That's a very long story 👼🏼. Initially it defaulted to `aufs`, then to different ones. It does do a little bit of this, depending on what it sees, it used one or the other (if not explicitely configured). But it's always better to be explicit in setting the `storageDriver` in `docker`.

**Comment by con-f-use** [comment]
https://github.com/NixOS/nixpkgs/pull/268979 <--- next try
